### DISCUSSION TITLE
Constrained Beam Search

### DISCUSSION THREAD

**Discussion by PolKul** [facebookresearch/ParlAI#3582] [Ideas]
Hello, 

I have been working on porting Constrained Beam Search algorithm from [Fairseq](https://github.com/pytorch/fairseq/pull/2402) library to ParlAI. You can check my fork [here](https://github.com/PolKul/ParlAI)

I was able to connect it with the Transformer agent. The demo can be run from tests/run_constrained_beam_search.py 

I still have several problems with the algorithm:

1. Currently I cannot use more than one constraint in the list (constraints parameter)
2. When using beam_context_block_ngram >0 it would output garbage in the generated text for the second and all consequent utterances

**Comment by mjpost** [comment]
Sorry, I don't know the ParlAI codebase or API at all, so I can't be of much assistance. Constraints work fine when used with fairseq-interactive, so there must be some problem with the interface.

**Reply by PolKul** [reply]
hi @mjpost, thanks for joining in. Well, I wasn't working with Fairseq either :) Anyway, the core logic is in the torch_generator_agent.py in the _generate() method. That is similar to the _generate() method from the sequence_generator.py from Fairseq. The logic is to take a batch, pass through an encoder, then decode and search (in ParlAI I have created a class ConstrainedBeamSearch where most of the Constrained Beam Search logic is implemented) iteratively until finished. 

You should be right, the problem is somewhere in the interface, maybe something to do with the outputs from the decoder (lprobs). It would be great if you could check.

**Reply by mjpost** [reply]
I'm a little confused. You posted a sequence of questions on the fairseq thread, and I am only now realizing that you were actually asking about your port of the code to ParlAI. Unfortunately, I don't have time to learn and debug a new codebase. My best suggestion is that you study the API in [interactive.py](https://github.com/pytorch/fairseq/blob/master/fairseq_cli/interactive.py) to see how to pass in multiple constraints. They have to be packed in a special tensor format.

**Comment by stephenroller** [comment]
Not sure but digging this energy y'all.

**Comment by PolKul** [comment]
@mjpost thats ok, ill try digging out myself. Actually I mentioned before that it was my custom ParlAI model. Anyway, I really appreciate the work you have done so far. 

@stephenroller i believe that making this algorithm work would benefit ParlAI greatly. Because without it we have little to none control over the generation. And we all know that generative models can go nuts if used without such control.

**Comment by PolKul** [comment]
just fixed the problem with multiple constraints. Works quite well now.

**Comment by stephenroller** [comment]
Will you consider submitting a PR?

**Reply by PolKul** [reply]
I've just created PR.
